cpp/src/arrow/gpu/cuda_context.cc

```diff
@@ -99,31 +100,52 @@ class CudaContext::CudaContextImpl {
     return Status::OK();
   }
 
+  Status CopyDeviceToDevice(void* dst, const void* src, int64_t nbytes) {
+    ContextSaver set_temporary(context_);
+    CU_RETURN_NOT_OK(cuMemcpyDtoD(reinterpret_cast<CUdeviceptr>(dst),
+                                  reinterpret_cast<const CUdeviceptr>(src),
+                                  static_cast<size_t>(nbytes)));
+    return Status::OK();
+  }
+
   Status Synchronize(void) {
     ContextSaver set_temporary(context_);
     CU_RETURN_NOT_OK(cuCtxSynchronize());
     return Status::OK();
   }
 
   Status Free(void* device_ptr, int64_t nbytes) {
-    CU_RETURN_NOT_OK(cuMemFree(reinterpret_cast<CUdeviceptr>(device_ptr)));
+    if (device_ptr != NULL)  // non 0-size buffer, see ExportIpcBuffer
+      CU_RETURN_NOT_OK(cuMemFree(reinterpret_cast<CUdeviceptr>(device_ptr)));
     bytes_allocated_ -= nbytes;
     return Status::OK();
   }
 
   Status ExportIpcBuffer(void* data, std::shared_ptr<CudaIpcMemHandle>* handle) {
     ContextSaver set_temporary(context_);
     CUipcMemHandle cu_handle;
-    CU_RETURN_NOT_OK(cuIpcGetMemHandle(&cu_handle, reinterpret_cast<CUdeviceptr>(data)));
+    if (data != NULL) {
+      CU_RETURN_NOT_OK(
+          cuIpcGetMemHandle(&cu_handle, reinterpret_cast<CUdeviceptr>(data)));
+    } else {
+      // 0-size buffers are marked by setting device pointer to 0. To
+      // make cu_handle aware of this, we zero cu_handle memory.
+      // Since in Linux sizeof(CUipcMemHandle) == 64, we cast handle
+      // value to int64_t:
+      reinterpret_cast<int64_t*>(&cu_handle)[0] = 0;
+    }
     *handle = std::shared_ptr<CudaIpcMemHandle>(new CudaIpcMemHandle(&cu_handle));
     return Status::OK();
   }
 
   Status OpenIpcBuffer(const CudaIpcMemHandle& ipc_handle, uint8_t** out) {
     auto handle = reinterpret_cast<const CUipcMemHandle*>(ipc_handle.handle());
-    CUdeviceptr data;
-    CU_RETURN_NOT_OK(
-        cuIpcOpenMemHandle(&data, *handle, CU_IPC_MEM_LAZY_ENABLE_PEER_ACCESS));
+    CUdeviceptr data = 0;
+    // If handle memory is zero, the IPC handle represents 0-size
+    // buffer. See also related comments in ExportIpcBuffer.
+    if (reinterpret_cast<const int64_t*>(handle)[0] != 0)
```

Comment:
The `reinterpret_cast<CUipcMemHandle>(0)` part gives invalid cast error.
The solution like in [here](https://stackoverflow.com/questions/6938219/how-to-check-whether-all-bytes-in-a-memory-block-are-zero) seems like an overkill.

Reply:
Another possibility is to stop using the "zero handle" hack and instead have a separate boolean indicating a zero-sized buffer. This probably means you need to include it in the IPC message structure.

Reply:
Whether a buffer is zero-sized, is determined by `size()` result. So, introducing a separate boolean for that might not be needed.

Reply:
For IPC, I think we need to store the buffer size (or just zero-size boolean) in CudaIpcMemHandle and include this to IPC message as @pitrou also suggested.

I would prefer to pass around buffer size than a boolean flag on `size==0`. Let me know if using the boolean flag would be more advantageous.

Reply:
I'm fine with passing the buffer size. It is also more informative.

Reply:
OK, now the buffer size is included in IPC buffer message. The message consists of `<memory size><ipc_handle>` if size > 0, otherwise it is just `<memory size>`.
Since the zero-sized buffers do not reference any memory, own_data is set false, and many memory managing methods are disabled. This simplified the patch.

I wonder, should we set own_data to false whenever buffer size is 0? Although we have related tests that pass ok.